### PR TITLE
add a flag to control the cap on images reported in node status

### PR DESCRIPTION
**Pull request by mtaufen**
While I normally try to avoid adding flags, this is a short term
scalability fix for v1.11, and there are other long-term solutions in
the works, so we shouldn't commit to this in the v1beta1 Kubelet config.
Flags are our escape hatch here.

```release-note
The maximum number of images the Kubelet will report in the Node status can now be controlled via the Kubelet's --node-status-max-images flag. The default (50) remains the same.
```

### CODE REVIEW

**Comment by mtaufen** [comment]
ah, looks like a test needs to be updated too

File: pkg/kubelet/kubelet_node_status_test.go [hunk 1]
```diff
@@ -1214,12 +1218,15 @@ func TestTryRegisterWithApiServer(t *testing.T) {
 }
 
 func TestUpdateNewNodeStatusTooLargeReservation(t *testing.T) {
```
Comment (dchen1107) [inline pkg/kubelet/kubelet_node_status_test.go]: Can you add one more test case with node-status-max-images=0?

**Comment by dchen1107** [comment]
LGTM except one comment for extra test coverage. Thanks!

cc/ @wojtek-t @davidopp @bgrant0607 @yujuhong

File: pkg/kubelet/kubelet_node_status_test.go (hunk 1 above)
Comment (mtaufen) [reply pkg/kubelet/kubelet_node_status_test.go]: added in TestUpdateNewNodeStatus

File: cmd/kubelet/app/options/options.go [hunk 2]
```diff
@@ -170,6 +170,10 @@ type KubeletFlags struct {
 	// bootstrapCheckpointPath is the path to the directory containing pod checkpoints to
 	// run on restore
 	BootstrapCheckpointPath string
+	// NodeStatusMaxImages caps the number of images reported in Node.Status.Images.
+	// The default value of 0 does not cap the number of images reported.
```
Comment (yujuhong) [inline cmd/kubelet/app/options/options.go]: Isn't the default value 50?

Comment (mtaufen) [reply cmd/kubelet/app/options/options.go]: Thanks, I'll fix

File: pkg/kubelet/kubelet_node_status_test.go [hunk 3]
```diff
@@ -86,8 +91,11 @@ func generateTestingImageList(count int) ([]kubecontainer.Image, []v1.ContainerI
 
 		expectedImageList = append(expectedImageList, apiImage)
 	}
-	// 3. only returns the top maxImagesInNodeStatus images in expectedImageList
-	return imageList, expectedImageList[0:maxImagesInNodeStatus]
+	// 3. only returns the top maxImages images in expectedImageList
+	if maxImages == 0 { // 0 means no limit
```
Comment (dashpole) [inline pkg/kubelet/kubelet_node_status_test.go]: Is `0 == unlimited` convention for k8s?  I seem to recall `-1 == unlimited` being more common, but you spend a lot more time around configuration than I do ;)

Comment (mtaufen) [reply pkg/kubelet/kubelet_node_status_test.go]: I guess we shouldn't block reporting none, if someone really wants to do that? Easy to change to -1.

Comment (mtaufen) [reply pkg/kubelet/kubelet_node_status_test.go]: PodPidsLimit has precedence for -1 being unlimited

Comment (mtaufen) [reply pkg/kubelet/kubelet_node_status_test.go]: QPS parameters use 0

Comment (dashpole) [reply pkg/kubelet/kubelet_node_status_test.go]: Sounds good

Comment (dashpole) [reply pkg/kubelet/kubelet_node_status_test.go]: hmmm

Comment (dashpole) [reply pkg/kubelet/kubelet_node_status_test.go]: I am fine with either then.  Up to you

File: pkg/kubelet/kubelet_node_status_test.go [hunk 4]
```diff
@@ -86,8 +91,11 @@ func generateTestingImageList(count int) ([]kubecontainer.Image, []v1.ContainerI
 
 		expectedImageList = append(expectedImageList, apiImage)
 	}
-	// 3. only returns the top maxImagesInNodeStatus images in expectedImageList
-	return imageList, expectedImageList[0:maxImagesInNodeStatus]
+	// 3. only returns the top maxImages images in expectedImageList
+	if maxImages == -1 { // 0 means no limit
```
Comment (dashpole) [inline pkg/kubelet/kubelet_node_status_test.go]: forgot to update comment

File: pkg/kubelet/kubelet_node_status_test.go (hunk 3 above)
Comment (mtaufen) [reply pkg/kubelet/kubelet_node_status_test.go]: Changed it to -1, which I think is more reasonable. 0 => limit to none is a natural thing to assume, IMO.

File: cmd/kubelet/app/options/options.go [hunk 5]
```diff
@@ -170,6 +170,9 @@ type KubeletFlags struct {
 	// bootstrapCheckpointPath is the path to the directory containing pod checkpoints to
 	// run on restore
 	BootstrapCheckpointPath string
+	// NodeStatusMaxImages caps the number of images reported in Node.Status.Images.
+	// This is an experimental, short-term flag to help with node scalability.
+	NodeStatusMaxImages int32
```
Comment (dashpole) [inline cmd/kubelet/app/options/options.go]: If it is experimental, should we add the `Experimental` prefix to the flag?  Or is that only for boolean flags (i.e. the thing is on or off).?

File: pkg/kubelet/kubelet_node_status_test.go (hunk 4 above)
Comment (mtaufen) [reply pkg/kubelet/kubelet_node_status_test.go]: thanks, updated

File: cmd/kubelet/app/options/options.go (hunk 5 above)
Comment (mtaufen) [reply cmd/kubelet/app/options/options.go]: I don't think it's necessary. We warn that it's alpha in the help text. In general, "experimental" isn't even a word that exists in the context of our deprecation policy (other than defining what "Alpha" means...), so we shouldn't use it in names, and we shouldn't force name changes when things graduate from alpha to beta in any case.

**Comment by wojtek-t** [comment]
LGTM

**Comment by yujuhong** [comment]
Uh...data race in the unit test
```
==================
WARNING: DATA RACE
Write at 0x00c4207dc7e4 by goroutine 37:
  k8s.io/kubernetes/pkg/kubelet.TestUpdateNewNodeStatus()
      pkg/kubelet/kubelet_node_status_test.go:466 +0x3df9
  testing.tRunner()
      GOROOT/src/testing/testing.go:777 +0x16d
...
```

**Comment by mtaufen** [comment]
It doesn't like my hack to hot-swap the config, lol.

**Comment by mtaufen** [comment]
refactored test to avoid the race

**Comment by wojtek-t** [comment]
@mtaufen - please rebase

**Comment by mtaufen** [comment]
re-added label after rebase

**Comment by mtaufen** [comment]
rebased

**Review by resouer — APPROVED** [review]

**Comment by bgrant0607** [comment]
Thanks.

@mtaufen This needs to have a release note, as it is a user-visible change in behavior.

**Comment by wojtek-t** [comment]
> @mtaufen This needs to have a release note, as it is a user-visible change in behavior.

@bgrant0607 - why? The default is still as it was. So I think it's not really changing the behavior (it just makes it configurable).

**Comment by mtaufen** [comment]
@bgrant0607 @wojtek-t Not a bit deal to add a relnote. Updated.

**Comment by bgrant0607** [comment]
@mtaufen @wojtek-t 

Adding a configuration option is an admin-relevant interface change.

https://github.com/kubernetes/community/issues/484

**Comment by jfbai** [comment]
Hi @mtaufen , is this flags going to be graduated to beta? or any other solution on this work?
We are going to apply this flag in our production, so I'd like to know the plan on this flag, thanks a lot. :)

**Comment by mtaufen** [comment]
The original plan was for the NodeLease feature to replace frequent Kubelet status updates, at which point having a configurable cap would be less of an issue. There was also some discussion about how there may be better/more compact ways to inform the scheduler what images are on a given node.

@wojtek-t, what's the progress on the NodeLease feature? Would it be ok to remove this flag at some point? Are there any concrete plans for moving to another way of reporting cached images?

**Comment by wojtek-t** [comment]
> @wojtek-t, what's the progress on the NodeLease feature? 

We're not graduating it to GA in 1.16, because we haven't heard about any usecases of using it seriously in production. But in all tests it looks great, so I expect it reaching GA in 1.17.

> Would it be ok to remove this flag at some point? 

Maybe?

> Are there any concrete plans for moving to another way of reporting cached images?

Not that I'm aware of.

**Comment by entropitor** [comment]
@wojtek-t @mtaufen What scalability issues are avoided with this flag and are they still a problem in newer versions of Kubernetes? 

I'm building an application on top of kubernetes and I'd need to increase this flag for it to work really well but I want to understand the consequences that this can have and I don't seem to be able to find this in the docs:

```
The maximum number of images to report in Node.Status.Images. If -1 is specified, no cap will be applied. (default 50)
```

**Comment by wojtek-t** [comment]
@entropitor - the more images you can have, the bigger the Node object is. Given that Nodes are periodically updates with heartbeats (every 10s up until 1.13, every 1m 1.14-1.16 (with NodeLease feature), and every 5m in 1.17+ [those are all default values], this may lead to significant load on apiserver and etcd on your cluster. This is becoming a problem on large clusters (at least hundreds of nodes) though.

**Pull request merged**